diff --git a/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWikiException.java b/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWikiException.java
index 1517f97..6580b00 100644
--- a/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWikiException.java
+++ b/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWikiException.java
@@ -120,6 +120,28 @@
     public static final int ERROR_XWIKI_STORE_HIBERNATE_CREATE_DATABASE = 3401;
     public static final int ERROR_XWIKI_STORE_HIBERNATE_DELETE_DATABASE = 3402;
     public static final int ERROR_XWIKI_STORE_HIBERNATE_CHECK_EXISTS_DATABASE = 3403;
+
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_DOC = 3501;
+    public static final int ERROR_XWIKI_STORE_JCR_READING_DOC = 3502;
+    public static final int ERROR_XWIKI_STORE_JCR_DELETING_DOC = 3503;
+    public static final int ERROR_XWIKI_STORE_JCR_CANNOT_DELETE_UNLOADED_DOC = 3504;
+    public static final int ERROR_XWIKI_STORE_JCR_READING_REVISIONS = 3503;
+    public static final int ERROR_XWIKI_STORE_JCR_READING_VERSION = 3504;
+    public static final int ERROR_XWIKI_STORE_JCR_UNEXISTANT_VERSION = 3505;
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_OBJECT = 3511;
+    public static final int ERROR_XWIKI_STORE_JCR_LOADING_OBJECT = 3512;
+    public static final int ERROR_XWIKI_STORE_JCR_DELETING_OBJECT = 3513;
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_CLASS = 3521;
+    public static final int ERROR_XWIKI_STORE_JCR_LOADING_CLASS = 3522;
+    public static final int ERROR_XWIKI_STORE_JCR_SEARCH = 3523;
+    public static final int ERROR_XWIKI_STORE_JCR_LOADING_ATTACHMENT = 3531;
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_ATTACHMENT = 3532;
+    public static final int ERROR_XWIKI_STORE_JCR_DELETING_ATTACHMENT = 3533;
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_ATTACHMENT_LIST = 3534;
+    public static final int ERROR_XWIKI_STORE_JCR_SEARCHING_ATTACHMENT = 3535;
+    public static final int ERROR_XWIKI_STORE_JCR_CHECK_EXISTS_DOC = 3536;
+    public static final int ERROR_XWIKI_STORE_JCR_SWITCH_DATABASE = 3601;
+    public static final int ERROR_XWIKI_STORE_JCR_CREATE_DATABASE = 3701;
     
     public static final int ERROR_XWIKI_RENDERING_VELOCITY_EXCEPTION = 4001;
     public static final int ERROR_XWIKI_RENDERING_GROOVY_EXCEPTION = 4002;
@@ -194,6 +216,17 @@
     public static final int ERROR_XWIKI_DIFF_ATTACHMENT_ERROR = 13026;
     public static final int ERROR_XWIKI_DIFF_XML_ERROR = 13027;
 
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_LOCK = 13106;
+    public static final int ERROR_XWIKI_STORE_JCR_LOADING_LOCK = 13107;
+    public static final int ERROR_XWIKI_STORE_JCR_DELETING_LOCK = 13108;
+    public static final int ERROR_XWIKI_STORE_JCR_INVALID_MAPPING = 13109;
+    public static final int ERROR_XWIKI_STORE_JCR_MAPPING_INJECTION_FAILED = 13110;
+    public static final int ERROR_XWIKI_STORE_JCR_LOADING_LINKS = 13111;
+    public static final int ERROR_XWIKI_STORE_JCR_SAVING_LINKS = 13112;
+    public static final int ERROR_XWIKI_STORE_JCR_DELETING_LINKS = 13113;
+    public static final int ERROR_XWIKI_STORE_JCR_LOADING_BACKLINKS = 13114;
+    public static final int ERROR_XWIKI_STORE_JCR_OTHER = 13130; // temporary
+
     public static final int ERROR_XWIKI_STORE_SEARCH_NOTIMPL = 13200;
 
     public static final int ERROR_XWIKI_GROOVY_COMPILE_FAILED = 14001;
